src/pip/_internal/cli/progress_bars.py

```diff
@@ -64,5 +101,16 @@ def get_download_progress_renderer(
     """
     if bar_type == "on":
         return functools.partial(_rich_progress_bar, bar_type=bar_type, size=size)
+    elif bar_type == "json":
+        # We don't want regular users to use this progress_bar type
+        # so only use if not a TTY
+        assert (
+            not sys.stdout.isatty()
+        ), 'The "json" progress_bar type should only be used inside subprocesses.'
```

Comment:
This needs to be a proper error, not an assert.

Reply:
Also, I'm not sure we should error if the JSON progress type is used from a console[^1] Maybe we should warn and use the standard progress bar? Something along the lines of "the JSON progress bar isn't appropriate for display directly to the user".

[^1]: Sorry if I'm changing my mind compared to what I said previously. I can't now find my previous comments to check what I'd said before.

Reply:
> Sorry if I'm changing my mind compared to what I said previously. I can't now find my previous comments to check what I'd said before.

You did, but it's fine. I think the warn and fallback to the regular progress bar is better anyway